### PR TITLE
Introduce a nested context for the deployer worker.

### PR DESCRIPTION
**Pull request by howbazaar**
This isn't yet hooked up into the deployer, but the nested context will
be hooked up to have the unit workers run in a nested dependency engine within
the deployer worker.

The Context interface is expanded to support the worker.Worker methods as
the NestedContext is a worker, but the SimpleContext has nothing running,
so its implementation are no-ops.

## QA steps

The unit tests coverage of the worker/deployer is all there is right now.

## Documentation changes

This is internal only for now.

### CODE REVIEW

File: worker/deployer/nested_test.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+func (s *NestedContextSuite) TestRecallUnit(c *gc.C) {
+	ctx := s.newContext(c)
+	unitName := "something/0"
+	err := ctx.DeployUnit(unitName, "password")
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Wait for unit to start.
+	s.workers.waitForStart(c, unitName)
+
+	err = ctx.RecallUnit(unitName)
```
Comment (wallyworld) [inline worker/deployer/nested_test.go:183]: Can we call this again to ensure it's idempotent

File: worker/deployer/unit_manifolds.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+	// Clock supplies timekeeping services to various workers.
+	Clock clock.Clock
+}
+
+// UnitManifolds returns a set of co-configured manifolds covering the various
+// responsibilities of nested unit agent.
+//
+// Thou Shalt Not Use String Literals In This Function. Or Else.
+func UnitManifolds(config UnitManifoldsConfig) dependency.Manifolds {
```
Comment (wallyworld) [inline worker/deployer/unit_manifolds.go:85]: Do we need this or can we reuse the existing unit agent manifolds?
I guess there will be subtle differences between what's needed for standalone vs embedded....

Comment (howbazaar) [reply worker/deployer/unit_manifolds.go:85]: We definitely need a different list. There are quite a few differences between them. Many of the unit agent manifolds have requirements around the upgrade worker having been run, but we don't even have that here. I felt it was better to have a clean implementation for the units here.

Originally I was planning to remove the old uniter code, but I have a feeling we may need it for the k8s side of things.

File: worker/deployer/nested.go [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+		"deployed": deployed.SortedValues(),
+		"units":    running,
+	}
+	if len(stopped) > 0 {
+		result["stopped"] = stopped.SortedValues()
+	}
+	return result
+}
+
+func (c *nestedContext) DeployUnit(unitName, initialPassword string) error {
```
[the remark this answers is not in the document]
Comment (howbazaar) [reply worker/deployer/nested.go:194]: Done.

File: worker/deployer/nested.go [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+)
+
+const (
+	deployedUnitsKey = "deployed-units"
+	stoppedUnitsKey  = "stopped-units"
+)
+
+// Logger represents a logger used by the context.
+type Logger interface {
+	Criticalf(string, ...interface{})
```
[the remark this answers is not in the document]
Comment (howbazaar) [reply worker/deployer/nested.go:37]: Removed.

File: worker/deployer/nested.go [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+		agent, err := NewUnitAgent(unitConfig)
+		if err != nil {
+			hasError = true
+			config.Logger.Errorf("unable to start unit %q: %v", u, err)
+			continue
+		}
+		context.units[u] = agent
+		if !stopped.Contains(u) {
+			if err := context.startUnit(u); err != nil {
+				return nil, errors.Annotatef(err, "issues starting workers for unit %q", u)
```
[the remark this answers is not in the document]
Comment (howbazaar) [reply worker/deployer/nested.go:150]: Updated to not have startUnit cause an error starting the context.

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**